### PR TITLE
Fixes issue #257 (intermittent failure to launch images due to ProcessWrapper causing error 998).

### PR DESCRIPTION
**Pull request by avasopht**
Fix to: #257, replacing ProcessWrapper with a more robust wrapper for win32\CreateProcess.

Error 998 is a common symptom of uninitialized structs, which I suspect caused the issue with ProcessWrapper. This error stopped immediately after making this fix, and it's been working without any issue for 6 weeks.

### CODE REVIEW

File: src/PharoLauncher-Process/PhLProcess.class.st [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+{ #category : #'*PharoLauncher-Process' }
+PhLProcess class >> createProcess: app command: cmd [ 
+	^ self createProcess: app command: cmd waits: true
+]
+
+{ #category : #'*PharoLauncher-Process' }
+PhLProcess class >> createProcess: app command: cmd waits: waits [
+	| startupInfo processInfo result cmdString kNoWindow |
+	kNoWindow := 16r08000000.
+	cmd ifNotNil: [ cmdString := ExternalAddress fromString: cmd ].
```
Comment (guillep) [inline src/PharoLauncher-Process/PhLProcess.class.st:26]: This only works for ascii strings

File: src/PharoLauncher-Process/PhLProcess.class.st [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+]
+
+{ #category : #'*PharoLauncher-Process' }
+PhLProcess >> closeHandle: handle [
+	^ self ffiCall: #(int CloseHandle(void * handle)) module: #Kernel32
+]
+
+{ #category : #'*PharoLauncher-Process' }
+PhLProcess >> createProcess: appName command: command processAttributes: procAttr threadAttributes: threadAttr inheritHandles: inherit creationFlags: flags environment: env directory: dir startup: startup processInformation: processInfo [
+	^ self ffiCall: #(int CreateProcessA (const char * appName, char * command, 0, 0, 0, ulong flags, 0, 0, ExternalAddress * startup, ExternalAddress * processInfo)) module: #Kernel32
```
Comment (guillep) [inline src/PharoLauncher-Process/PhLProcess.class.st:83]: Any chance to make it work with CreateProcessW? If you check the package `System-Platforms-Windows` there are already some classes to do windows wide string marshalling

File: src/PharoLauncher-Process/PhLProcess.class.st [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+{ #category : #'*PharoLauncher-Process' }
+PhLProcess >> memSet: ptr value: value size: size [
+	^ self ffiCall: #(ExternalAddress * memset(ExternalAddress * ptr, int value, size_t size))
+]
+
+{ #category : #'*PharoLauncher-Process' }
+PhLProcess >> waitForSingleObject: handle milliSeconds: ms [
+	^ self ffiCall: #( uint WaitForSingleObject (void * handle, uint ms)) module: #Kernel32
+]
+
```
Comment (guillep) [inline src/PharoLauncher-Process/PhLProcess.class.st:100]: This file is really strange: it has all extension methods that should be in an extension file, and the extension file (`PhlProcess.extension.st`) is repeating all this code all over the place.

Did you use Iceberg to commit this (and did you select all changes)?

**Review by guillep — CHANGES_REQUESTED** [review]
I like this change, I think it goes in a nice direction, and it would make making the 64bit launcher working more easily.

I have some comments on the PR however.

@demarey do we have tests on these things? It should not be complicated to write some basic tests on launching images... Otherwise this kind of change requires lots of manual testing...

File: src/PharoLauncher-Process/PhLProcess.class.st (hunk 3 above)
Comment (avasopht) [reply src/PharoLauncher-Process/PhLProcess.class.st:100]: Hi @guillep, I originally wrote those methods as an extension before moving them into PhLProcess. Seems I forgot to create the protocols. I'll update that.

File: src/PharoLauncher-Process/PhLProcess.class.st (hunk 2 above)
Comment (avasopht) [reply src/PharoLauncher-Process/PhLProcess.class.st:83]: Yes that's fine. I'll update it to support wide strings asap.

File: src/PharoLauncher-Core/PhLProcessWrapper.class.st [hunk 4]
```diff
@@ -40,27 +40,20 @@ PhLProcessWrapper class >> encode: aCommand [
...
 
 { #category : #'execution - public' }
 PhLProcessWrapper class >> waitForCommand: aCommand [
 	| encodedCommand |
 	encodedCommand := self encode: aCommand.
 	^ self new
-		process: ((Smalltalk os isWindows and: [ Smalltalk includesKey: #ProcessWrapper ])
-			ifTrue: [ self waitForWindowsCommand: encodedCommand ]
+		process: (Smalltalk os isWindows
+			ifTrue: [ self windowsCommand: encodedCommand ]
```
Comment (demarey) [inline src/PharoLauncher-Core/PhLProcessWrapper.class.st:56]: should'nt it be self waitForWindowsCommand: here?

**Comment by demarey** [comment]
Hi Keldon, 

Thanks for this pull request. It is true that the fork of processes on Windows is not reliable enough. Using Windows API looks to be the right way.

@guillep There are tests for image version determination using process forks: https://github.com/pharo-project/pharo-launcher/blob/development/src/PharoLauncher-Tests-Functional/PhLVirtualMachineManagerFunctionalTest.class.st#L42.
There are also a few tests on forking processes:
https://github.com/pharo-project/pharo-launcher/blob/development/src/PharoLauncher-Tests-Functional/PhLProcessWrapperTest.class.st
But the point is that test are only checked Linux for now.

@avasopht It would be great if you could use the CreateProcessW function. We recently had issues because of the encoding. For example, in french, non-ASCII characters are frequent in names, so you can easily find non-ASCII characters for the HOME directory, part of the path to images and vms.

**Comment by demarey** [comment]
@avasopht Thanks a lot for this PR. It gave me the opportunity to dive a bit in Windows process creation and take time to package a library that works on the Unicode API and having a similar API as OSSubprocess: https://github.com/pharo-contributions/OSWinSubprocess. If it is ok for you, I will close this PR. Pharo Launcher will use this library if feedbacks are good. There is still work to do on this library but it is a good start.

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/pharo-contributions/OSWinSubprocess">pharo-contributions/OSWinSubprocess</a></strong></div><div>Forking Windows System Processes from within Pharo Language  - pharo-contributions/OSWinSubprocess</div></blockquote>

**Comment by avasopht** [comment]
> @avasopht Thanks a lot for this PR. It gave me the opportunity to dive a bit in Windows process creation and take time to package a library that works on the Unicode API and having a similar API as OSSubprocess: https://github.com/pharo-contributions/OSWinSubprocess. If it is ok for you, I will close this PR. Pharo Launcher will use this library if feedbacks are good. There is still work to do on this library but it is a good start.
> 
> >  GitHub**[pharo-contributions/OSWinSubprocess](https://github.com/pharo-contributions/OSWinSubprocess)**Forking Windows System Processes from within Pharo Language  - pharo-contributions/OSWinSubprocess

Brilliant!

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/pharo-contributions/OSWinSubprocess">pharo-contributions/OSWinSubprocess</a></strong></div><div>Forking Windows System Processes from within Pharo Language  - pharo-contributions/OSWinSubprocess</div></blockquote>

**Pull request closed without merging**